lib/chef/resource.rb

```diff
@@ -772,29 +779,88 @@ def provider=(arg)
     def self.property(name, type=NOT_PASSED, **options)
       name = name.to_sym
 
-      if type != NOT_PASSED
-        if options[:is]
-          options[:is] = ([ type ] + [ options[:is] ]).flatten(1)
-        else
-          options[:is] = type
-        end
-      end
+      local_properties = properties(false)
+      type = properties[name] if type == NOT_PASSED && properties[name]
+      local_properties[name] = property_type(type, name: name, declared_in: self, **options)
+      local_properties[name].emit_dsl
+    end
 
-      define_method(name) do |value=NOT_PASSED|
-        set_or_return(name, value, options)
-      end
-      define_method("#{name}=") do |value|
-        set_or_return(name, value, options)
-      end
+    #
+    # Create a reusable property type that can be used in multiple properties
+    # in different resources.
+    #
+    # @param type [Object,Array<Object>] The type(s) of this property.
+    #   If present, this is prepended to the `is` validation option.
+    #   If this is a Chef::Property, `specialize` is called on it to create the
+    #   new property instead of prepending to `is`.
+    # @param options [Hash<Symbol,Object>] Validation options. see #property for
+    #   the list of options.
+    #
+    # @example Bare property_type
+    #   property_type()
+    #
+    # @example With just a type
+    #   property_type(String)
+    #
+    # @example With just options
+    #   property_type(default: 'hi')
+    #
+    # @example With type and options
+    #   property_type(String, default: 'hi')
+    #
+    def self.property_type(type=NOT_PASSED, **options)
```

Comment:
Or to be more specific, it would probably be better for API stability to make this `(**options)` since we are expecting external code it monkey patch this method. Type can just become `options[:type]` in the code calling this.